content/source/docs/enterprise-beta/api/modules.html.md

```diff
@@ -0,0 +1,189 @@
+      "published_at": "2017-11-30T00:53:06.966753Z",
+      "downloads": 0,
+      "verified": false
+    },
+    {
+      "id": "skierkowski-v2/subnet/aws/0.0.1",
+      "owner": "",
+      "namespace": "skierkowski-v2",
+      "name": "subnet",
+      "version": "0.0.1",
+      "provider": "aws",
+      "description": "",
+      "source": "https://bitbucket-server.hashicorp.engineering/projects/SKI/repos/terraform-aws-subnet/browse",
+      "published_at": "2017-11-30T17:38:01.600527Z",
+      "downloads": 0,
+      "verified": false
+    }
+  ]
+}
+```
+
+## Publish a Module
+
+This endpoint can be used to publish a new module to the registry. The publishing process will ingress the configuration from the repository and parse the various fields (e.g. input variables, output variables, readme) from the repository.
+
+| Method | Path           |
+| :----- | :------------- |
+| POST | /registry-modules |
+
+### Parameters
+
+- `linkable-repo-id` (`string: <required>`) - Specifies the repository to be used to ingress the configuration, this is in the format `<PROJECT_KEY>/<REPO>`
+- `o-auth-token-id` (`string: <requires>`) - Specifies the VCS Connection (OAuth Conection + Token) to use as identified. This ID can be obtained from the [o-auth-tokens](./o-auth-tokens.html) endpoint.
+
+
+### Sample Payload
+They payload must include `ingress-trigger-attributes`; however, those fields **MUST** be set to the default values provided.
+
+```json
+{
+  "data": {
+    "attributes": {
+      "linkable-repo-id":"SKI/terraform-aws-instance",
+      "o-auth-token-id":"1",
+      "ingress-trigger-attributes": {
+        "branch":"",
+        "default-branch":true,
+        "vcs-root-path":""
+      }
+    },
+    "type":"registry-modules"
+  }
+}
+```
+
+### Sample Request
+
+```shell
+curl \
+  --header "Authorization: Bearer $ATLAS_TOKEN" \
+  --header "Content-Type: application/vnd.api+json" \
+  --request POST \
+  --data @payload.json \
+  https://atlas.hashicorp.com/api/v2/registry-modules
+```
+
+### Sample Response
+
+```json
+{
+  "data": {
+    "id": "mod-1JqHG3j71bwoukuX",
+    "type": "registry-modules",
+    "attributes": {
+      "name": "instance",
+      "provider": "aws",
+      "status": "pending",
+      "version-statuses": [],
+      "created-at": "2017-11-30T00:00:52.386Z",
+      "updated-at": "2017-11-30T00:00:52.386Z",
+      "permissions": {
+        "can-delete": true,
+        "can-resync": true,
+        "can-retry": true
+      }
+    },
+    "relationships": {
+      "organization": {
+        "data": {
+          "id": "org-QpXoEnULx3r2r1CA",
+          "type": "organizations"
+        }
+      }
+    },
+    "links": {
+      "self": "/api/v2/registry-modules/show/skierkowski-v2/instance/aws"
+    }
+  }
+}
+```
+
+## Delete a module
+
+This set of endpoint can be used to delete a single version, all versions for a given module and provider, or all versions for all providers for a given module.
```

Comment:
We should add that deletes cascade probably. I've tried explaining this in RFCs and internal docs abotu 3 times now and not found a non-clumsy way to put it so you can try and do better but the technical behaviour is:

 1. If you delete a specific provider, and that provider was the only provider for that module, the whole module is deleted.
 2. If you delete a specific version, and that version was the only version for that provider, the whole provider record is deleted, and if the provider was the only provider for that module the module is deleted too.

Reply:
How about this:

> These endpoints can delete a single version for a provider, a single provider (and all its versions) for a module, or a whole module. If the requested deletion would leave a provider with no versions or a module with no providers, the empty items will be automatically deleted as well.

Reply:
Sounds great :)